include/boost/gil/histogram.hpp

```diff
@@ -0,0 +1,247 @@
+                    "Improper container type for signed images.");
+
+    using channel_t = typename channel_type<SrcView>::type;
+    using pixel_t = pixel<channel_t, gray_layout_t>;
+
+    hist.resize(std::numeric_limits<channel_t>::max() + 1);
+
+    for_each_pixel(color_converted_view<pixel_t>(srcview), [&hist](pixel_t const& p) {
+        ++hist[p];
+    });
+}
+
+/// \ingroup Histogram
+/// \brief Use for indexable and resizable sequence containers
+///
+/// This overload is the same as the above except that it does not produce the
+/// accumulate effect for histogram bin values. It returns a new container of
+/// size large enough to hold entire range of corresponding pixel value in the input image.
+///
+template
+<
+    typename Container,
+    typename SrcView,
+    typename std::enable_if<
+                detail::is_indexable<Container>::value &&
+                detail::is_resizable<Container>::value, int>::type = 0
+>
+inline Container image_histogram(SrcView const& srcview)
+{
+    gil_function_requires<ImageViewConcept<SrcView>>();
+    static_assert(std::is_arithmetic<typename Container::value_type>::value,
+                    "Improper container type for images.");
+    static_assert(std::is_unsigned<typename channel_type<SrcView>::type>::value,
+                    "Improper container type for signed images.");
+
+    using channel_t = typename channel_type<SrcView>::type;
+    using pixel_t = pixel<channel_t, gray_layout_t>;
+
+    Container hist(std::numeric_limits<channel_t>::max() + 1);
+    
+    for_each_pixel(color_converted_view<pixel_t>(srcview), [&hist](pixel_t const& p) {
+        ++hist[p];
+    });
+    return hist;
+}
+
+/// \ingroup Histogram
+/// \brief Use for indexable and non-resizable sequence containers 
+///
+/// This overload differs from the first, it gives a partial histogram over the image
+/// for non-resizable sequence containers. No overload of accumulate type is provided in
+/// this case. Example container would be std::array.
+///
+template
+<
+    typename Container,
+    typename SrcView,
+    typename std::enable_if <
+                detail::is_indexable<Container>::value &&
+                !detail::is_resizable<Container>::value &&
+                std::is_arithmetic<typename Container::value_type>::value, int>::type = 0
+>
+inline void image_histogram(SrcView const& srcview, Container &hist)
+{
+    gil_function_requires<ImageViewConcept<SrcView>>();
+    static_assert(std::is_arithmetic<typename Container::value_type>::value,
+                    "Improper container type for images.");
+    static_assert(std::is_unsigned<typename channel_type<SrcView>::type>::value,
+                    "Improper container type for signed images.");
+
+    using channel_t = typename channel_type<SrcView>::type;
+    using pixel_t = pixel<channel_t, gray_layout_t>;
+    
+    for_each_pixel(color_converted_view<pixel_t>(srcview), [&hist](pixel_t const& p) {
+        if (p < hist.size()) ++hist[p];
```

Comment:
instead shouldn't this create equal size bins and store the data according to the bins!?

Reply:
I was not sure how we could convey that piece of information to the end user i.e. specify what the bin size they should use when accessing the values.